src/org/javarosa/core/reference/ReferenceManager.java

```diff
@@ -203,28 +191,14 @@ public void clearSession() {
         logger.debug("cleared all session translators");
     }
 
+    /**
+     * Returns the first deriving factory in the three ReferenceFactory collections
+     */
     private ReferenceFactory derivingRoot(String uri) throws InvalidReferenceException {
-
-        //First, try any/all roots which are put in the temporary session stack
-        for (RootTranslator root : sessionTranslators) {
-            if (root.derives(uri)) {
-                return root;
-            }
-        }
-
-        //Now, try any/all roots referenced at runtime.
-        for (RootTranslator root : translators) {
-            if (root.derives(uri)) {
-                return root;
-            }
-        }
-
-        //Now try all of the raw connectors available
-        for (ReferenceFactory root : factories) {
-            if (root.derives(uri)) {
-                return root;
-            }
-        }
+        for (List<? extends ReferenceFactory> rfs : Arrays.asList(sessionTranslators, translators, factories))
```

Comment:
You could do this:
```java
private ReferenceFactory derivingRoot(String uri) throws InvalidReferenceException {
    for (ReferenceFactory rf : getAllReferenceFactoriesInOrder())
        if (rf.derives(uri))
            return rf;

    throw new InvalidReferenceException(getPrettyPrintException(uri), uri);
}

private List<ReferenceFactory> getAllReferenceFactoriesInOrder() {
    List<ReferenceFactory> refFactories = new ArrayList<>();
    refFactories.addAll(sessionTranslators);
    refFactories.addAll(translators);
    refFactories.addAll(factories);
    return refFactories;
}
```

What do you think?

Reply:
No problem! :dagger: 

I can't help to not try to fix design problems. Those subtypes are leaking because that hierarchy is not sealed tight. The code I'm suggesting is more verbose but deals with that design flaw. It also flattens the lists, which is kind of nice too.

I won't cry if you don't change it :)

Reply:
Thank you for not crying. 

Sometime I’d like to understand the design flaw this would deal with.

Reply:
(I don't know why a dagger was showing up in my last comment!)

This is something super common not only in the ODK codebase, but in many other projects as well, and I think we can blame Java and how Java has been historically teached for that.

The flaw is that the hierarchy formed by `ReferenceFactory` (supertype) and its subtypes `RootTranslator`, `PrefixedRootFactory`, `ResourceReferenceFactory`, ..., is leaking the subtypes.

You can see this leak in many places. Related to this change we're talking about:
- You can't change the type of the `ReferenceManager` `translators` and `sessionTranslators` to a `List<ReferenceFactory>` because it's coupled with a method on `RootTranslator`.
- Because of this, you can't directly get a `List<List<ReferenceFactory>>` as a result of an `Array.asList(sessionTranslators, translators, factories)`.

If you're a SOLID design fan, this is breaking the L (Liskov's Substitution Principle), but in a more general way, this is wrong because it's a misuse of inheritance.

Inheritance is a tool we use to implement polymorphism, and here we don't have polymorphism, which is weird because we could benefit from it (the `for` block we're discussing would benefit from it).

In a polymorphic type, you can't have the subtypes leaked outside the type hierarchy (no one can be coupled to them, even implicitly, like with the `List<? extends ReferenceFactory>`).

If this is not a polymorphic type, this is telling me that:
- Maybe we have used inheritance for code reuse, which would be much simpler to get by composing stuff together
- Maybe we have used inheritance systematically without taking into account the impact this has in evolving software and the cost of introducing new changes.

This is even worse because the `ReferenceFactory` is an extension point that clients use to adapt JR to their particular environment. It's really confusing that JR is coupled both to the supertype (OK) and the subtypes (wrong).

Since this is all across the codebase, it's not super critical to fix in this particular instance but you know the Boy Scout's rule: leave the place cleaner than you found it.

Reply:
This is very educational!